### PR TITLE
[Ray] task based shuffle for ray

### PR DESCRIPTION
**Pull request by chaokunyang**
## What do these changes do?
This PR implements pull-based shuffle for ray task backend, which will be the basis for later push-based shuffle for ray. Main changes include:

- [x] Introduce `ShuffleFetchType` which contains `FETCH_BY_KEY`/`FETCH_BY_INDEX`. If `FETCH_BY_INDEX` is used, `source_keys/source_idxes/source_mappers` will be skipped. In this way the meta and scheduling cost for supervisor will be much smaller, about 1000~10000x smaller for big shuffle. This will make mars able to schedule large-scale shuffle, even PB-scale data shuffle.
- [x] add n_mappers/n_reducers to `FetchShuffle`, and skip `source_keys/source_idxes/source_mappers` when `shuffle_fetch_type` is `FETCH_BY_INDEX`. 
- [x] Operands compatibility: make mars operands compatible with `FETCH_BY_INDEX`. 
  - [x] Mappers may produce undefinite output nums: `TensorReshape`/`TensorBincount`
  - [x] Mappers may produce shuffle blocks greater than num_reducer: `PSRSAlign`.
- [x] Introduce a new subtask dependency resolve mechanism, which alse remove `data_keys` for subtask execution on ray
- [x] Add `ShuffleManager` for shuffle execution on ray
- [x] Skip recording shuffle operands meta in supervisor
- [x] Support all shuffle operands on ray

Before this PR, for a shuffle with 100000 partitions, every reducer subtask meta will took about 16M memory, which make mars to handle only less than 1000 partitions, i.e. less than 1 T data.

With this PR, every reducer subtask took only 2Kb memory, which make mars can handle 8000_000 partitions. If every partition has 1Gb data, this PR will enable mars handle 8PB data shuffle.

#2916
#3039
#2112

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/deploy/oscar/local.py
Comment (fyrestone) [inline mars/deploy/oscar/local.py:157]: We don't needs this because the execution config merging will update the backend.

File: mars/deploy/oscar/tests/test_ray_dag.py
Comment (fyrestone) [inline mars/deploy/oscar/tests/test_ray_dag.py:134]: Why not just run the `test_local.py::test_execute_describe`?

File: mars/deploy/utils.py
Comment (fyrestone) [inline mars/deploy/utils.py:161]: Don't add the `backend` here, `backend` is for execution. You can add the `shuffle_type` to the execution config is it is related to the `backend`.

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:118]: Any tests for the input keys that matched all the cases of shuffle operands?

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:129]: Maybe we can add a decorator to the `execute_subtask` to log the exceptions.

File: mars/tensor/reshape/reshape.py
Comment (fyrestone) [inline mars/tensor/reshape/reshape.py:465]: This adds more costs to the Mars backend, can we move this to Ray executor?

File: mars/deploy/oscar/tests/test_ray_dag.py
Comment (chaokunyang) [reply mars/deploy/oscar/tests/test_ray_dag.py:134]: we added `df.groupby` in this tests case, `test_execute_describe` doesn't contains `groupby`.

File: mars/deploy/utils.py
Comment (chaokunyang) [reply mars/deploy/utils.py:161]: I think it's needed.
image
If we don't add this, and the execution backend is `ray`, `config["task"]["execution_config"]["backend"]` will still be `mars` which is not consistent.

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:118]: Yes, will add test cases to cover shuffle operands. we definitely need do this, otherwise the shuffle implementation will be easily broke

File: mars/tensor/reshape/reshape.py
Comment (chaokunyang) [reply mars/tensor/reshape/reshape.py:465]: Actually the cost is minimal. Mars shuffle fetch will filter out all `None` values.

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:129]: I added a decorator at the first, but I removed it, cause I felt it kind of complex. 
This is my decorator definition:
```
def log_exception(func, *msg_args, ex_logger=None):
    import inspect
    if msg_args:
        msg_tpl, *msg_args = msg_args
    else:
        msg_tpl, msg_args = None, []
    if inspect.iscoroutinefunction(func):
        async def new_func(*args, **kwargs):
            try:
                return await func(*args, **kwargs)
            except Exception:
                (ex_logger or logger).exception(msg_tpl, *msg_args)
                raise
    else:
        def new_func(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except Exception:
                (ex_logger or logger).exception(msg_tpl, *msg_args)
                raise
    return new_func
```

Do we need add this?

File: mars/deploy/oscar/local.py
Comment (chaokunyang) [reply mars/deploy/oscar/local.py:157]: It didn't update the config in `config["task"]["execution_config"]["backend"]`

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:129]: emm, `log_exception` seems reusable, will change to decorator

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:119]: There is no easy way for a unit test cases cover all shuffle operands. I added assert in `execute_subtask`. The runtime  cost is minimal, I think it's acceptable.

File: mars/deploy/oscar/tests/test_ray_dag.py
Comment (fyrestone) [reply mars/deploy/oscar/tests/test_ray_dag.py:134]: Then, just extends the `test_execute_describe` case or add a new case? We are trying to get both of the backends to run the same test cases.

File: mars/deploy/utils.py
Comment (fyrestone) [reply mars/deploy/utils.py:161]: Here updates the config, we can switch the backend by passing `backend="ray"` to the `new_session` and `new_test_session`.

image

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [reply mars/services/task/execution/ray/executor.py:119]: Can you add some assertion to the `CheckedSubtaskProcessor` to make sure the map keys you generated is exactly the same with Mars original logic.

File: mars/deploy/utils.py
Comment (chaokunyang) [reply mars/deploy/utils.py:161]: Got it, I passed `shuffle_type` from `TaskProcessor` to `GraphAnalyzer` directly.

File: mars/deploy/oscar/tests/test_ray_dag.py
Comment (chaokunyang) [reply mars/deploy/oscar/tests/test_ray_dag.py:134]: Split it into `test_execute_describe` and `test_shuffle` test cases

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:119]: The keys will generated only when `PushShuffle` is used. Mars dones't build `PushShuffle` into subtask graph.

File: mars/tensor/reshape/reshape.py
Comment (chaokunyang) [reply mars/tensor/reshape/reshape.py:465]: Yes, `None` values will be put into storage in Mars backend. But shuffle mapper meta are not recorded in mars, so the cost will be negligible？

**Comment by zhongchun** [comment]
Is this implementation the solution proposed in MEP 0001?  Is `PushShuffle` the `FetchShuffleByIndex` mentioned in MEP 001?

**Comment by chaokunyang** [comment]
> Is this implementation the solution proposed in MEP 0001? Is `PushShuffle` the `FetchShuffleByIndex` mentioned in MEP 001?

Yes, this is the implementation for  MEP 0001. I've removed `PushShuffle` and add the properties into `FetchShuffle`

File: mars/utils.py
Comment (fyrestone) [inline mars/utils.py:563]: Why not put these lines to `elif shuffle_fetch_type == ShuffleFetchType.FETCH_BY_KEY`? It seems that when `shuffle_fetch_type == ShuffleFetchType.FETCH_BY_INDEX`, the `source_keys, source_idxes, source_mappers` are None.

File: mars/utils.py
Comment (fyrestone) [inline mars/utils.py:593]: Is this branch `if isinstance(chunk_op, ShuffleProxy):` necessary? I noticed that the `ShuffleProxy` type is checked in the analyzer, if we return `build_fetch_shuffle(chunk)`, the `n_reducers` will be None, `assert n_reducers > 0, n_reducers` will fail.

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:133]: Can you dump the subtask graph dots if error occurred? We want the success execution branch as fast as possible.

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:651]: All the pure dependants are yield and executed when tiling (the iterative tiling), can you provide some cases that the execution is out of order?

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:161]: The `iter_indep` order is guaranteed, please transfer the `input_keys` for non shuffle subtasks.

File: mars/utils.py
Comment (chaokunyang) [reply mars/utils.py:593]: Good catch, we should remove this and assert chunk.op is not ShuffleProxy

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:161]: I omitted `input_keys ` on purpose, so the meta for task spec will be small nad serialization cost is small too.

File: mars/utils.py
Comment (chaokunyang) [reply mars/utils.py]: Fixed

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:651]: The comments is stale, I'll remove it.

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:133]: Fixed

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [reply mars/services/task/execution/ray/executor.py:159]: My point is the order of `iter_indep` is not guaranteed through serialize / deserialze. If the `input_keys` is omitted here, then we should add more checks to keep the order when serializing.

File: mars/services/task/execution/ray/executor.py
Comment (fyrestone) [inline mars/services/task/execution/ray/executor.py:170]: `subtask_chunk_graph.to_dot()` every execute is very slow. We can pass a lambda instead of string formatter to the `log_exception_wrapper`.

Comment (fyrestone) [reply mars/services/task/execution/ray/executor.py:170]: Also, `log_exception_wrapper` every loop is very slow, too. Please wrap the entire loop or just wrap the `execute_subtask`.

File: mars/core/graph/core.pyx
Comment (fyrestone) [inline mars/core/graph/core.pyx:367]: Is the `to_dot` changes are necessary for ray shuffle? If not, it's better to move them to another PR.

File: mars/tensor/random/core.py
Comment (fyrestone) [inline mars/tensor/random/core.py:317]: When the `tup` is None?

Comment (chaokunyang) [reply mars/tensor/random/core.py:317]: when serializing `random_state`, if passed random state is  None, the `_on_serialize_random_state` just return None. In such situation,  the `tup` is `None`. I'm not sure why `random_state` is None, but I think it's a bug in previous and without this fix, `GroupBySample` shuffle on ray will fail. Mars tests won't fail just because it runs in test session which no serialziation happend. The exception is soemthing like follows and we can submmit another PR to fix it:
```
E         File "mars/serialization/core.pyx", line 924, in mars.serialization.core.deserialize
E           raise exc_value
E         File "mars/serialization/core.pyx", line 888, in mars.serialization.core.deserialize
E           deserialized = _deserial_single(
E         File "mars/serialization/core.pyx", line 802, in mars.serialization.core._deserial_single
E           res = serializer.deserial(serialized[_COMMON_HEADER_LEN:], context, subs)
E         File "mars/serialization/core.pyx", line 103, in mars.serialization.core.Serializer.deserial
E           cpdef deserial(self, tuple serialized, dict context, list subs):
E         File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/core/operand/base.py", line 338, in deserial
E           operand: Operand = super().deserial(serialized, context, subs)
E         File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/serialization/serializables/core.py", line 238, in deserial
E           self._set_field_value(obj, field, value)
E         File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/serialization/serializables/core.py", line 220, in _set_field_value
E           field.set(obj, field.on_deserialize(value))
E         File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/tensor/random/core.py", line 317, in _on_deserialize_random_state
E           rs.set_state(tup)
E         File "mtrand.pyx", line 357, in numpy.random.mtrand.RandomState.set_state
E       TypeError: state must be a dict or a tuple.
```

Comment (chaokunyang) [reply mars/tensor/random/core.py:317]: Fixed it in #3149, I'll merge master after it's merged

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:159]: How about add `sort_by` parameter to `iter_indep` oor add a `iter_indep_sorted` API to `DirectedGraph`. I did a rough tests, sort chunks is much faster tghen serialization. And if we omit `input_keys`, the task spec is small, we can also cache more task spec for lineage reconstruction 
image

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py]: reverted to print exception directly

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py]: Changed to return None for pure_depdend_keys instead of coresponding ObjectRef

File: mars/services/ray_utils.py
Comment (fyrestone) [inline mars/services/ray_utils.py:23]: The original `_get_ray_executor` accepts no arguments, so it's OK to set the maxsize to None. The `_ray_export_once` accepts variety functions or classes, the `maxsize=None` will have potential memleak.

If the export once is not used very often, can we just leave it as it is and not introduce a new util?

File: mars/services/task/execution/ray/shuffle.py
Comment (fyrestone) [inline mars/services/task/execution/ray/shuffle.py:77]: If the `reducer_subtasks` is less than `n_reducers` as the comment, the missing value in the `filled_reducers` will be a number not None. The comment is not match with the code logic.

File: mars/services/task/execution/ray/shuffle.py
Comment (fyrestone) [inline mars/services/task/execution/ray/shuffle.py:66]: If the item of `sorted_filled_reducer_subtasks` may be None, is it should be skipped?

File: mars/services/task/execution/ray/shuffle.py
Comment (fyrestone) [inline mars/services/task/execution/ray/shuffle.py:141]: When goes to `else`?

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:172]: It depends on the keys size, the sort algorithm has `O(n)log(n)` time complexity. If the keys are small, sort is much faster than serialization.
image
And the input keys in subtask chunk graph is less than 5 normally so the cost is ignorable.

I agree with you that sort should be used sparely. If we can avoid sort, the code will be more clean. If we serializing dict by two list, the order will be fixed. But sort is more explict and customizing may should be put in another PR if we really need it?

File: mars/core/entity/tileables.py
Comment (chaokunyang) [reply mars/core/entity/tileables.py:90]: Many chunks are not contained in tileable.chunks, such as `df.describe`.  We can't get all generated chunks here.
image

File: mars/core/operand/fetch.py
Comment (chaokunyang) [reply mars/core/operand/fetch.py:59]: Aggred, `ReferenceField` is better

File: mars/core/operand/shuffle.py
Comment (chaokunyang) [reply mars/core/operand/shuffle.py:133]: Just for `fetch_by_index`

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:231]: Good suggestion, this helps make code more cleaner.

File: mars/services/task/execution/ray/executor.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/executor.py:187]: I prefer catch exception per loop.
image
is more cleaner for me compared to:
image

File: mars/services/task/execution/ray/shuffle.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/shuffle.py:48]: Good suggestion, I'll submit another PR to add `n_reducers` and `n_mappers` to proxy

Comment (chaokunyang) [reply mars/services/task/execution/ray/shuffle.py:48]: added `n_reducers/n_mappers` in https://github.com/mars-project/mars/pull/3160

File: mars/services/ray_utils.py
Comment (chaokunyang) [reply mars/services/ray_utils.py:23]: 1. As name `once` in `_ray_export_once` shows it export a `class/function` only once, so it won't be called frequently, and memleak won't happen.
2. Since memleak won't happen and the code ara repeted. why not just make it as an util?

File: mars/services/task/execution/ray/shuffle.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/shuffle.py:141]: If the `subtask` isn't a mapper sutbask, it will go to `else` branch.

File: mars/core/entity/tileables.py
Comment (qinxuye) [reply mars/core/entity/tileables.py:90]: Just check the `mars/mars/services/task/supervisor/tests/task_preprocessor.py` to see if it's possible to integrate the check into it, because the check logic is sealed in the tests module instead of putting it directly into the normal code which looks really ugly.

File: mars/services/task/execution/ray/shuffle.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/shuffle.py]: Good catch, Fixed it

File: mars/core/entity/tileables.py
Comment (chaokunyang) [reply mars/core/entity/tileables.py:90]: Actually I tried to add the checks to `CheckedTaskPreprocessor` at first, but the `CheckedTaskPreprocessor` only have `ChunkGraph` as inputs. Since some reducers are not contained in the chunk graph. Add checks here won't cover all cases. I agree with you add the checks in normal code is not elegant. But seems we can't find a place to get all tiled reducers. The checks here just to capture operands implementation error.

File: mars/services/ray_utils.py
Comment (fyrestone) [reply mars/services/ray_utils.py:23]: If we make it as a util, then it may be used some where by somebody. We can't assume that the `_ray_export_once` only exports one or two `class / function`. My suggestion is mentioned as above,

- Not make it as a util (fastest).
- Make it as a util and remove `maxsize=None` to avoid potential memleak.

File: mars/services/task/execution/ray/shuffle.py
Comment (fyrestone) [reply mars/services/task/execution/ray/shuffle.py:64]: As above mentioned, the item of `sorted_filled_reducer_subtasks` may be None, so it's possible to update a `{None: (shuffle_index, reducer_ordinal)}` to the `self. reducer_indices`. Is that expected?

File: mars/services/ray_utils.py
Comment (chaokunyang) [reply mars/services/ray_utils.py]: Reverted

File: mars/core/entity/tileables.py
Comment (chaokunyang) [reply mars/core/entity/tileables.py:90]: The checks here just to capture operands implementation error. The normal codepath doesn't need this.

File: mars/services/task/execution/ray/shuffle.py
Comment (fyrestone) [reply mars/services/task/execution/ray/shuffle.py:139]: This is a bit clearer:
```python
index = self.mapper_indices.get(subtask) or self.reducer_indices.get(subtask)
if index is None:
    raise Exception("The {subtask} should be a mapper or a reducer.")
else:
    shuffle_index, _ = index
```

**Comment by chaokunyang** [comment]
Since this PR introduce `build_fetch_shuffle` which doesn't exist on master, some asv benchmarks will fail
image

File: mars/tests/test_utils.py
Comment (fyrestone) [inline mars/tests/test_utils.py:675]: It seems that the `log_exception_wrapper` code is not used. Can you remove it and the related tests from this PR? Then you can add them back in some PR when they are needed.

File: mars/utils.py
Comment (fyrestone) [inline mars/utils.py:581]: Is the fetch op only has one output? I noticed that the `[params]` is passed to kws.

File: mars/tensor/reshape/reshape.py
Comment (fyrestone) [inline mars/tensor/reshape/reshape.py:472]:
```suggestion
            *(range(dim_chunk_cnt) for dim_chunk_cnt in dim_chunk_counts)
```

File: mars/services/task/execution/ray/shuffle.py
Comment (fyrestone) [reply mars/services/task/execution/ray/shuffle.py:139]: I noticed that the `get_n_reducers` is called only when the subtask is mapper, so is the `else` needed?

File: mars/core/entity/tileables.py
Comment (chaokunyang) [reply mars/core/entity/tileables.py]: Moved the checks to `CheckedTaskPreprocessor`

File: mars/services/task/execution/ray/shuffle.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/shuffle.py:64]: They are already updated to `self.reducer_indices`.

File: mars/tests/test_utils.py
Comment (chaokunyang) [reply mars/tests/test_utils.py:675]: reverted

File: mars/utils.py
Comment (chaokunyang) [reply mars/utils.py:581]: `FetchShuffle` only have one output which is reduce op. But is this related to this PR?

File: mars/services/task/execution/ray/shuffle.py
Comment (chaokunyang) [reply mars/services/task/execution/ray/shuffle.py:139]: `else` can be reserved for use in future

**Review by fyrestone — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**
